packages/gateway/src/manager.ts

```diff
@@ -54,6 +55,163 @@ export function createGatewayManager(options: CreateGatewayManagerOptions): Gate
+          },
+        })
+
+        if (gateway.preferSnakeCase) {
+          shard.forwardToBot = async (payload) => {
+            if (payload.t === 'READY') {
+              await gateway.resharding.updateGuildsShardId(
+                (payload.d as DiscordReady).guilds.map((g) => g.id),
+                shardId,
+              )
+            }
+
+            options.events?.message?.(shard, payload)
+          }
+        }
+
+        gateway.resharding.shards.set(shardId, shard)
+
+        const bucket = gateway.buckets.get(shardId % gateway.connection.sessionStartLimit.maxConcurrency)
+        if (!bucket) return
+
+        return await new Promise((resolve) => {
+          // Mark that we are making an identify request so another is not made.
+          bucket.identifyRequests.push(resolve)
+          logger.debug(`[Gateway] identifying shard #(${shardId}).`)
+          // This will trigger identify and when READY is received it will resolve the above request.
+          shard?.identify().then(async () => {
+            // Tell the manager that this shard is online
+            return await gateway.resharding.shardIsPending(shard)
+          })
+        })
+      },
+      async updateGuildsShardId(guildIds, shardId) {
+        logger.warn(
+          `[Resharding] Updating the following guild ids shard to #${shardId}: ${guildIds.join(', ')}. Override this function to update your cache if you need to.`,
+        )
+      },
+      async shardIsPending(shard) {
+        // Save this in pending at the moment, until all shards are online
+        gateway.resharding.pendingShards.set(shard.id, shard)
+        logger.warn(`[Resharding] Shard #${shard.id} is now pending`)
+        // Check if all shards are now online.
+        if (gateway.lastShardId - gateway.firstShardId + 1 > gateway.resharding.pendingShards.size) return
+        logger.warn(`[Resharding] All shards are now online.`)
+
+        // New shards start processing events
+        for (const shard of gateway.resharding.pendingShards.values()) {
+          shard.events.message = options.events?.message
+        }
+
+        // Old shards stop processing events
+        for (const shard of gateway.shards.values()) {
+          const oldHandler = shard.events.message
+
+          // Change with spread operator to not affect new shards, as changing anything on shard.events will directly change options.events, which changes new shards' events
+          shard.events = {
+            ...shard.events,
+            message: async function (_, message) {
+              // Member checks need to continue but others can stop
+              if (message.t !== 'GUILD_MEMBERS_CHUNK') return
+              // Process only the chunking events
+              oldHandler?.(shard, message)
+            },
+          }
+        }
+
+        logger.warn(`[Resharding] Shutting down old shards.`)
+        // Close old shards
+        await gateway.shutdown(ShardSocketCloseCodes.Resharded, 'Resharded!')
```

Comment:
Should we shut down all shards immediately? What if there are pending guild member chunks that are still incoming? Should we maybe have a function that requests permission from the user and waits until it is granted for specific shards and then shut them down? (either individually or wait to get ok signal for all shards and then do `gateway.shutdown`?)

Reply:
Why can't the user just override the shutdown method for that?

Reply:
Although I think your right but I don't think that should block this or any longer no?

Reply:
It can be added separately or it can be done by user

Reply:
True, that's fair